proof/crefine/RISCV64/IpcCancel_C.thy

```diff
@@ -2948,198 +3052,261 @@ lemma cancelIPC_ccorres1:
-               apply (case_tac "tcbState tcb", simp_all add: is_cap_fault_def)[1]
-              apply ceqv
-             apply ccorres_remove_UNIV_guard
-             apply (rule ccorres_move_array_assertion_tcb_ctes)
-             apply (rule_tac P="tcb_at' thread" in ccorres_cross_over_guard)
-             apply (simp add: getThreadReplySlot_def)
-             apply ctac
-               apply (simp only: liftM_def bind_assoc return_bind del: Collect_const)
-               apply (rule ccorres_pre_getCTE)
-               apply (rename_tac slot slot' cte)
-               apply (rule ccorres_move_c_guard_cte)
-               apply (rule_tac xf'=ret__unsigned_longlong_' and val="mdbNext (cteMDBNode cte)"
-                         and R="cte_wp_at' ((=) cte) slot and invs'"
-                          in ccorres_symb_exec_r_known_rv_UNIV[where R'=UNIV])
-                  apply vcg
-                  apply (clarsimp simp: cte_wp_at_ctes_of)
-                  apply (erule(1) cmap_relationE1[OF cmap_relation_cte])
-                  apply (clarsimp simp: typ_heap_simps)
-                  apply (clarsimp simp: ccte_relation_def map_option_Some_eq2)
-                 apply ceqv
-                apply csymbr
-                apply (rule ccorres_Cond_rhs)
-                 apply (simp add: nullPointer_def when_def)
-                 apply (rule ccorres_symb_exec_l[OF _ _ _ empty_fail_stateAssert])
-                   apply (rule ccorres_symb_exec_r)
-                     apply (ctac add: cteDeleteOne_ccorres[where w1="scast cap_reply_cap"])
-                    apply vcg
-                   apply (rule conseqPre, vcg, clarsimp simp: rf_sr_def
-                       gs_set_assn_Delete_cstate_relation[unfolded o_def])
-                  apply (wp | simp)+
-                apply (rule ccorres_return_Skip)
-               apply (simp add: guard_is_UNIV_def ghost_assertion_data_get_def
-                                ghost_assertion_data_set_def cap_tag_defs)
-              apply (simp add: locateSlot_conv, wp)
+               apply (rule_tac
+                        Q'="\<lambda>rv s.
+                               st_tcb_at' ((=) (Structures_H.thread_state.BlockedOnReceive bo bicg ro)) thread s
+                               \<and> tcb_at' thread s
+                               \<and> valid_objs' s
+                               \<and> weak_sch_act_wf (ksSchedulerAction s) s
+                               \<and> no_0_obj' s
+                               \<and> pspace_aligned' s
+                               \<and> pspace_distinct' s
+                               \<and> pspace_bounded' s
+                               \<and> valid_tcbs' s
+                               \<and> (ro = None \<longrightarrow> True)
+                               \<and> (\<forall>x. ro = Some x \<longrightarrow> obj_at' (\<lambda>reply. replyTCB reply = Some thread) x s)"
+                        in hoare_post_imp)
```

Comment:
This kind of explicit goal rephrasing is a sign that backwards reasoning is breaking down for some reason. Usually that's not a good sign, but it is at times unavoidable. Any thoughts on what happened?

Reply:
This mess was an artifact of the approach I took to figuring out how to do this proof. It's unnecessary, and I've removed it and moved the following bit to the end of the proof where it belongs.

What happened is that I was having trouble with a lhs Hoare triple goal with an option split in the postcondition. Not having discovered `hoare_case_option_wp2` yet, I tried a bunch of different things and ended up down this goal-rephrasing path in my attempt to make incremental progress. Once I finally did discover `hoare_case_option_wp2` and was able move on with the rest of the proof, and again during my subsequent synthesis+streamlining passes, I missed realizing the goal rephrasing was unnecessary.

Reply:
There are things done during figuring out proofs that rarely make it to the final product, but are perfectly reasonable methods to go about it. Hand-holding instantiations, splitting up the proof at a "known" juncture to understand which of the facts is actually necessary, and so on. This looked like one of those cases :)